// controllers/download.js
var db = require('controllers/database');
var mongoose = require('mongoose');
var downloads = mongoose.model('downloads');
var demands = mongoose.model('demands');
var users = mongoose.model('users');

exports.getDownloadEnd = function(country,city,callback) {
	downloads.find({country:country,city:city,status:true},function(err,data){
		if(data.length != 0){
			callback({'res':true,'data':data});
		}else{
			callback({'res':false});
		}
	});
}

exports.addDownload = function(name,size,token,username,city,country,callback) {
	var idObject = mongoose.Types.ObjectId();
	var newdownload = new downloads({
		_id: idObject,
		name: name,
		size: size,
		date: new Date(),
		status: "pending",
		token: token,
	 	username: username,
		city: city,
		country: country,
	});
	var newdemand = new demands({
		idService: idObject,
		service: "Download",
		name: name,
		token: token,
		username: username,
		main: true,
		status: "pending",
		notify: false,
		pt: 0,
		index: 0
	});

	newdownload.save(function (err){
		if(err) {
			callback({'res':false, 'response':"Error Downloads, you have not added a new download."});
		}else{
			newdemand.save(function (err){
				if(err) {
					callback({'res':false, 'response':"Error Demands, you have not added a new download."});
				}else{
					callback({'res':true, 'response':"Successfully Downloads, you have added a new download."});
				}
			});
		}
	});
}
